File: Other/Review/Phase 9/Review4/Part1/bfs-dfs/bfs-dfs.go
package main

import "fmt"

type TestNode struct {
	value     int
	neighbors []*TestNode
}

func BFS(root *TestNode) int {
	queue := []*TestNode{}
	depth := 0

	if root == nil {
		return depth
	}

	queue = append(queue, root)

	for len(queue) > 0 {

		size := len(queue)

		for i := 0; i < size; i++ {
			current := queue[0]
			queue = queue[1:]

			fmt.Println(current.value)

			for _, neighbor := range current.neighbors {
				queue = append(queue, neighbor)
			}
		}
		depth++
	}

	return depth
}

func DFS_Recursion(root *TestNode) {

	if root == nil {
		return
	}

	var recursion func(node *TestNode)
	recursion = func(node *TestNode) {
		if node == nil {
			return
		}

		fmt.Println(node.value)

		for _, neighbor := range node.neighbors {
			recursion(neighbor)
		}
	}

	recursion(root)
}

func DFS_Iteration(root *TestNode) {
	stack := []*TestNode{}

	if root == nil {
		return
	}

	stack = append(stack, root)

	for len(stack) > 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		fmt.Println(current.value)

		for _, neighbor := range current.neighbors {
			stack = append(stack, neighbor)
		}
	}
}

// test
func main() {
	root := TestNode{
		value: 1,
		neighbors: []*TestNode{
			{
				value: 2,
				neighbors: []*TestNode{
					{
						value:     4,
						neighbors: []*TestNode{},
					},
					{
						value:     5,
						neighbors: []*TestNode{},
					},
				},
			},
			{
				value: 3,
				neighbors: []*TestNode{
					{
						value: 6,
						neighbors: []*TestNode{
							{
								value:     7,
								neighbors: []*TestNode{},
							},
						},
					},
				},
			},
		},
	}

	// BFS
	res := BFS(&root)
	fmt.Println("layer: ", res)

	// DFS Recursion
	// fmt.Println("------")
	DFS_Recursion(&root)

	// // DFS Iteration
	// fmt.Println("------")
	DFS_Iteration(&root)
}

File: Other/Review/Stage 2/Review 2/bfs-dfs/bfs-dfs.swift
import Foundation

protocol TestNode {
    var value: Any { get }
    var neighbors: [TestNode] { get }
}

func BFS(root: TestNode?) -> Int {
    guard let root = root else {
        return 0
    }

    var queue: [TestNode] = [root]
    var depth = 0

    while !queue.isEmpty {
        let size = queue.count

        for _ in 0..<size {
            let current = queue.removeFirst()
            print(current.value)

            for neighbor in current.neighbors {
                queue.append(neighbor)
            }
        }

        depth += 1
    }

    return depth
}

func DFS_Iteration(root: TestNode?) {
    guard let root = root else {
        return
    }

    var stack: [TestNode] = [root]

    while !stack.isEmpty {
        let current = stack.removeLast()
        print(current.value)

        for neighbor in current.neighbors {
            stack.append(neighbor)
        }
    }
}

func DFS_Recursion(root: TestNode?) {
    guard let root = root else {
        return
    }

    func recursive(node: TestNode) {
        print(node.value)

        for neighbor in node.neighbors {
            recursive(node: neighbor)
        }
    }

    recursive(node: root)
}


struct Node: TestNode {
    var value: Any
    var neighbors: [TestNode]
}

// test
let root: TestNode = Node(value: 1, neighbors: [
    Node(value: 2, neighbors: [
        Node(value: 4, neighbors: []),
        Node(value: 5, neighbors: [])
    ]),
    Node(value: 3, neighbors: [
        Node(value: 6, neighbors: [
            Node(value: 7, neighbors: [])
        ])
    ])
])

// BFS
let res = BFS(root: root)
print("layer", res)

// DFS Recursion
print("------")
DFS_Recursion(root: root)

// DFS Iteration
print("------")
DFS_Iteration(root: root)